src/librustc_middle/ty/util.rs

```diff
@@ -778,6 +778,27 @@ impl<'tcx> ty::TyS<'tcx> {
         }
     }
 
+    /// Returns `true` if this type implements both `PartialStructuralEq` and `StructuralEq`.
+    ///
+    /// These marker traits are implemented along with `PartialEq` and `Eq` when implementations
+    /// for those traits are derived. They indicate that two values of this type are equal if all
+    /// of their fields are equal. A quirk of the marker traits is that their implementation is
+    /// never conditional on generic parameter bounds. For that reason, this helper function does
+    /// not take a `ParamEnv`.
+    ///
+    /// This function is "shallow" because it may return `true` for a type whose fields are not
+    /// `StructuralEq`. You will need to use a type visitor if you want to know whether a call to
+    /// `PartialEq::eq` will proceed structurally for a given type.
```

Comment:
```suggestion
    /// `PartialEq::eq` will proceed structurally for a given type.
    /// 
    /// Note: This function is "precise" when called on an ADT type; 
    /// that is, for an input ADT type, this will return `true` if
    /// *and only if* the input ADT type implements both
    /// `PartialStructuralEq` and `StructuralEq`. For non-ADT's, it
    /// does not provide that strong a guarantee.
```